youtube_dl/extractor/medaltv.py

```diff
@@ -0,0 +1,123 @@
+# coding: utf-8
+
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    try_get,
+    float_or_none,
+    int_or_none
+)
+
+
+class MedalTVIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?medal\.tv/clips/(?P<id>[0-9]+)'
+    _TESTS = [{
+        'url': 'https://medal.tv/clips/34934644/3Is9zyGMoBMr',
+        'md5': '7b07b064331b1cf9e8e5c52a06ae68fa',
+        'info_dict': {
+            'id': '34934644',
+            'ext': 'mp4',
+            'title': 'Quad Cold',
+            'description': 'Medal,https://medal.tv/desktop/',
+            'uploader': 'MowgliSB',
+            'timestamp': 1603165266,
+            'upload_date': '20201020',
+            'uploader_id': 10619174,
+        }
+    }, {
+        'url': 'https://medal.tv/clips/36787208',
+        'md5': 'b6dc76b78195fff0b4f8bf4a33ec2148',
+        'info_dict': {
+            'id': '36787208',
+            'ext': 'mp4',
+            'title': 'u tk me i tk u bigger',
+            'description': 'Medal,https://medal.tv/desktop/',
+            'uploader': 'Mimicc',
+            'timestamp': 1605580939,
+            'upload_date': '20201117',
+            'uploader_id': 5156321,
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        hydration_data = self._search_regex(r'<script>.*hydrationData\s*=\s*({.+?})\s*</script>', webpage, 'hydration data', default='{}')
+        parsed = self._parse_json(hydration_data, video_id)
+
+        # This is necessary because the id of the author is not known in advance.
+        # Will not raise an issue if no profile can be found as this is optional.
+        author_info = try_get(parsed.get('profiles'), lambda x: list(x.values())[0], dict) or {}
+        author_url = 'https://medal.tv/users/{}'.format(author_info.get('id')) if author_info.get('id') else None
+
+        clip_info = try_get(parsed.get('clips'), lambda x: x.get(video_id), dict) or {}
+        if(not clip_info):
+            raise ExtractorError('Could not find video information.', video_id=video_id)
+
+        if('error' in clip_info):
```

Comment:
error into a common variable and reuse it.
it would be better to check for error only after not been able extracted mandatory information(`video_id`, `title` and `formats`).
```suggestion
        error = clip_info.get('error')
        if error:
```

Reply:
When you say "it would be better to check for error only after not been able extracted mandatory information(video_id, title and formats).", I believe I have already done this?

```python
        clip_info = try_get(parsed, lambda x: x['clips'][video_id], dict) or {}
        if(not clip_info):
            raise ExtractorError('Could not find video information.', video_id=video_id)

        error = clip_info.get('error')
        if(error):
            if(error == 404):
                raise ExtractorError('That clip does not exist.',
                                     expected=True, video_id=video_id)
            else:
                raise ExtractorError('An unknown error occurred ({0}).'.format(error),
                                     video_id=video_id)
```

I check for error `error = clip_info.get('error')` only after not being able to extract mandatory information: `if(not clip_info):`. `clip_info` contains the mandatory information, so if that does not exist, then the whole process should break.

Reply:
even if `clip_info` exist, it's possibe that there is no format there(empty `contentUrl*` values).

Reply:
Okay, I see what you mean, but the JSON returned for an error (when a clip does not exist) differs in structure from a normal clip:

WORKING CLIP:
```json
{
   "clips":{
      "34934644":{
         "contentId":34934644,
         "contentTitle":"Quad Cold",
         "contentDescription":"Medal,https://medal.tv/desktop/",
         "sourceWidth":1920,
         "sourceHeight":1080,
         "videoLengthSeconds":23,
         "contentUrl1080p":"https://cdn.medal.tv/10619174/1080p-34934644.mp4",
         "contentUrl720p":"https://cdn.medal.tv/10619174/720p-34934644.mp4",
         "contentUrl480p":"https://cdn.medal.tv/10619174/720p-34934644.mp4",
         "contentUrl360p":"https://cdn.medal.tv/10619174/360p-34934644.mp4",
         "contentUrl240p":"https://cdn.medal.tv/10619174/360p-34934644.mp4",
         "contentUrl144p":"https://cdn.medal.tv/10619174/360p-34934644.mp4",
         "thumbnail1080p":"https://cdn.medal.tv/10619174/thumbnail-34934644-720p.jpg",
         "thumbnail720p":"https://cdn.medal.tv/10619174/thumbnail-34934644-720p.jpg",
         "thumbnail480p":"https://cdn.medal.tv/10619174/thumbnail-34934644-720p.jpg",
         "thumbnail360p":"https://cdn.medal.tv/10619174/thumbnail-34934644-360p.jpg",
         "thumbnail240p":"https://cdn.medal.tv/10619174/thumbnail-34934644-360p.jpg",
         "thumbnail144p":"https://cdn.medal.tv/10619174/thumbnail-34934644-360p.jpg",
         "contentUrl":"https://cdn.medal.tv/source/10619174/1603165173-medaltvcallofdutyrblackopscoldwar20201019203926.mp4",
         "socialMediaVideo":"https://cdn.medal.tv/10619174/share-34934644.mp4",
         "created":1603165266000,
         "likes":80,
         "views":8702,
         "comments":18,
         "id":34934644
      }
   },
   "profiles":{
      "10619174":{
         "id":10619174,
         "userId":10619174,
         "displayName":"MowgliSB",
         "userName":"MowgliSB"
      }
   }
}
```

CLIP NOT FOUND:
```json
{
   "clips":{
      "12345":{
         "id":"12345",
         "error":404
      }
   },
   "profiles":{
      "undefined":{
         "status":"None",
         "error":{
            "message":"invalid json response body at http://api-lb.api/users/undefined reason: Unexpected end of JSON input",
            "type":"invalid-json"
         }
      }
   }
}
```

So, if `error` is in `clip_info`, then it automatically is invalid. This is why I check for `error` in `clip_info` before anything else. Because if `error` is in `clip_info`, there will never be format information.

You say that
> even if clip_info exist, it's possibe that there is no format there(empty contentUrl* values).

But if `clip_info` exists (i.e. `parsed['clips'][video_id]` exists) and `parsed['clips'][video_id]['error']` exists, then there will not be a `contentUrl` value.

Are you saying that I must check if at least one format is returned? E.g. in the case when there is `clip_info`, no `clip_info['error']` and no `clip_info['contentUrl']`? If so, surely that will be covered in the case when there are no formats retrieved? 
```ERROR: No video formats found!; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.``` Although it has never happened to me, if it does happen, youtube-dl will handle it correctly.

Reply:
it's also to cover the other possiblility, both `error` and `contentUrl*` exist, the response may change for different errors.
so in general it's prefered to check the existance of formats before raising an error(ex: [YahooIE](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/yahoo.py#L278-L279), [GoogleDriveIE](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/googledrive.py#L249), etc...).

Reply:
Okay I see! Thanks for explaining 😃👍